src/client/gui/gui_cmd.cpp

```diff
@@ -0,0 +1,299 @@
+
+    auto reply = list_future.result();
+
+    for (auto it = instances_entries.cbegin(); it != instances_entries.cend(); ++it)
+    {
+        auto instance = std::find_if(reply.instances().cbegin(), reply.instances().cend(),
+                                     [it](const ListVMInstance& instance) { return it->first == instance.name(); });
+
+        if (instance == reply.instances().cend())
+        {
+            instances_to_remove.push_back(it->first);
+        }
+    }
+
+    for (const auto instance : instances_to_remove)
+    {
+        instances_entries.erase(instance);
+    }
+
+    for (const auto& instance : reply.instances())
+    {
+        auto name = instance.name();
+        auto state = QString::fromStdString(mp::format::status_string_for(instance.instance_status()));
+
+        auto it = instances_entries.find(name);
+
+        if (it != instances_entries.end())
+        {
+            auto instance_state = it->second.state;
+            if (state == "DELETED")
+            {
+                instances_entries.erase(name);
+            }
+            else if (instance_state != state)
+            {
+                auto& instance_menu = instances_entries.at(name).menu;
+
+                instance_menu->setTitle(QString("%1 (%2)").arg(QString::fromStdString(name)).arg(state));
+                instance_menu->clear();
+                set_menu_actions_for(name, state);
+                instances_entries[name].state = state;
+            }
+
+            continue;
+        }
+
+        if (state != "DELETED")
+        {
+            instances_entries[name].menu =
+                std::make_unique<QMenu>(QString("%1 (%2)").arg(QString::fromStdString(name)).arg(state));
+            set_menu_actions_for(name, state);
+            instances_entries[name].state = state;
+        }
+    }
+
+    if (instances_entries.empty())
+    {
+        about_separator->setVisible(false);
+    }
+    else
+    {
+        about_separator->setVisible(true);
+    }
+}
+
+void cmd::GuiCmd::update_about_menu()
+{
+    auto reply = version_future.result();
+
+    about_client_version.setText("multipass version: " + QString::fromStdString(multipass::version_string));
+    about_daemon_version.setText("multipassd version: " + QString::fromStdString(reply.version()));
+
+    if (update_available(reply.update_info()))
+    {
+        update_action.setWhatsThis(QString::fromStdString(reply.update_info().url()));
+        tray_icon_menu.insertAction(about_menu.menuAction(), &update_action);
+    }
+    else
+    {
+        tray_icon_menu.removeAction(&update_action);
+    }
+}
+
+void cmd::GuiCmd::create_menu()
+{
+    tray_icon.setContextMenu(&tray_icon_menu);
+
+    tray_icon.setIcon(QIcon{":images/multipass-icon.png"});
+
+    QObject::connect(&list_watcher, &QFutureWatcher<ListReply>::finished, this, &GuiCmd::update_menu);
+
+    QObject::connect(&menu_update_timer, &QTimer::timeout, [this]() { initiate_menu_layout(); });
```

Comment:
For safety, could you make "this" the third argument, and the lambda the fourth? That'll ensure if the GuiCmd object is deleted, the connection will be broken so the lambda won't fire. It might save a crash on shutdown.